### PR TITLE
Finish 3D Atten,Random Erasing Loss,Group Deconvolution;

### PR DESCRIPTION
**Pull request by b762927**
1.Finish 3D Atten,Random Erasing Loss,Group Deconvolution;
2.The resnet.py has been modifified to transplant the 3D Atten;(Here is modified)
3. Atten3D.py is the core of this code;
4. train.py add the ER LOSS(Random Erasing Loss);(Here is modified)
5. cvt_cifar_100.py add the function of Random Erasing of the transforms.py to process the dataset;(Here is modified)
6. src,function,modules folders are the official implementation of the deformable conv;
7. build.py and make.sh is to bulid the deformable conv;
8. transform.py is the Random Erasing official code;
9. the model_resnet can be used by the ImageNet(224*244)，but it can modify to be better

### CODE REVIEW

File: datasets/cvt_cifar_100.py [hunk 1]
```diff
@@ -4,20 +4,22 @@
...
 
 dataset_path = '/home/yuange/dataset/cifar-100'
+#TODO:是加入pytorch的ER还是使用论文源码的ER(代码中的：transforms.py)
 
 train_data = torchvision.datasets.CIFAR100(
     root=dataset_path,
     train=True,
     transform=torchvision.transforms.Compose([
             torchvision.transforms.ToTensor(),
+            tran.RandomErasing(probability=0.5, sh=0.4, r1=0.3),#使用ER loss再加入
```
Comment (ygtxr1997) [inline datasets/cvt_cifar_100.py:22]: cvt_cifar_100.py 是将下载好的 CIFAR-100 数据集读取并打包成 mxnet 格式的数据集。
所以数据增广不应该放在这里。
而应该放在 datasets/dataset_arcface.py 的 MXCifarTrainDataset 或者 MXCifarTestDataset 下面，这个类会被 train.py 调用。

File: backbones/Atten3D.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+#TODO:这里得引入下面模块才能使用分组可变形，然后num_deformable_groups究竟设为多少？(类别数还是？)
+#from modules import DeformConv
+
+num_deformable_groups=2#这里是可变形分组卷积的组数设置
```
Comment (ygtxr1997) [inline backbones/Atten3D.py:7]: 赋值表达式以及类似运算式的编写规范是 xxx(空格)=(空格)yyy。
代码的注释规范是在代码后面加两个空格，并且＃符号后面要加一个空格，例如：
num_deformable_groups(空格)=(空格)2(空格)(空格)#(空格)这里是可变形......

File: backbones/Atten3D.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        y = torch.split(x, 1, dim=1)  # 在第一维度(通道)分离
+        list = []
+        for i in y:
+            i = i.view(-1, self.inSize * self.inSize)
+            i=self.fc(i)
+            list.append(i)
+
+        z = torch.stack(list, dim=1)  # 把前面分裂后的y在第二维度拼接
+        return z
+
```
Comment (ygtxr1997) [inline backbones/Atten3D.py:25]: class前面要空两行

File: backbones/model_resnet.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
+import torch
```
Comment (ygtxr1997) [inline backbones/model_resnet.py:1]: 这个代码看上去像是 CBAM 的官方实现，如果是的话，可以直接在 backbones/resnet.py 上面修改，而不用另外再写一份 resnet。

File: backbones/model_resnet.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+class ResNet(nn.Module):
+    def __init__(self, block, layers,  network_type, num_classes):
+        self.inplanes = 64#输入通道(3->64)
+        super(ResNet, self).__init__()
+        self.network_type = network_type
+        # different model config between ImageNet and CIFAR 
+        if network_type == "ImageNet":
+            self.conv1 = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False)#stage1的卷积，输出为112*112*64
+            self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)#stage1的最大池化，输出为56*56*64
+            self.avgpool = nn.AvgPool2d(7)#均值池化是在最后一层用的
```
Comment (ygtxr1997) [inline backbones/model_resnet.py:101]: 可以直接用 nn.AdaptiveAvgPool2d 来代替 nn.AvgPool2d 这种写法。
在 backbones/resnet.py 中使用的就是 nn.AdaptiveAvgPool2d。

File: backbones/resnet.py [hunk 6]
```diff
@@ -10,8 +10,8 @@
 
 import torch
 import torch.nn as nn
-
-
+from torchinfo import summary
```
Comment (ygtxr1997) [inline backbones/resnet.py:13]: 这个库要小心使用，因为它会忽略某些层的计算。可参考调用时的 warn info。
如果某些层的flops和params被忽略，就需要我们自行计算某些模块的flops和params。

File: backbones/resnet.py [hunk 7]
```diff
@@ -10,8 +10,8 @@
...
+from backbones.Atten3D import  *
```
Comment (ygtxr1997) [inline backbones/resnet.py:14]: __all__ = ...... 前面也要空两行

File: backbones/resnet.py [hunk 8]
```diff
@@ -99,11 +99,34 @@ def __init__(self, block, num_block, num_classes=100,
...
         #we use a different inputsize than the original paper
         #so conv2_x's stride is 1
-        self.conv2_x = self._make_layer(block, 64, num_block[0], 1)
-        self.conv3_x = self._make_layer(block, 128, num_block[1], 2)
-        self.conv4_x = self._make_layer(block, 256, num_block[2], 2)
-        self.conv5_x = self._make_layer(block, 512, num_block[3], 2)
-        self.avg_pool = nn.AdaptiveAvgPool2d((1, 1))
+
+        # --------------------------stage 2------------------------------
+        self.Atten2 = Atten3D(64, 8, 12, 2)  # 输出为32*32*64
```
Comment (ygtxr1997) [inline backbones/resnet.py:117]: 这种暴力侵入式修改会影响原版resnet的调用。
比较好的修改方式是，在 __init__ 中添加一个 default 参数，例如 `en_atten3d=False`。
并且按照如下方式初始化：
```python
self.Atten2 = Atten3D(64, 8, 12, 2) if self.en_atten3d else nn.Identity()
```
但是我发现如果要把 3d attention 加进原版 resnet 的话，会有很多修改，需要很多 if 语句，比较难看。
所以我觉得可以在 backbones/ 目录下再创建一个 resnet_ours.py，把 resnet.py 拷贝过去然后再把我们的 3d attention 和 deformable 加上去。
resnet_ours.py 的最后部分仿照 resnet.py 定义 resnet18, resnet34, resnet50, resnet101 等等，但是不要用同样的名字，可参考如下命名方式：
```python
class ResNet(self, block, num_block, 
                 num_classes=100,
                 fp16=False,
                 en_3d=False,
                 en_defor=False,
                 en_erloss=False):
    ....

def resnet18_3d(**kwargs):
    ....

def resnet18_de(**kwargs):
    ....

def resnet18_3d_de(**kwargs):
    ....

```

File: backbones/resnet.py [hunk 9]
```diff
@@ -133,16 +158,56 @@ def _make_layer(self, block, out_channels, num_blocks, stride):
...
+            output = output * atten  # dot product
             output = self.conv5_x(output)
+
+            #--------------pool+fc---------------
             output = self.avg_pool(output)
             output = output.view(output.size(0), -1)
         output = self.fc(output.float() if self.fp16 else output)
 
-        return output
+        return output#,deAtten2,deAtten3,deAtten4,deAtten5  ##for the ER LOSS
```
Comment (ygtxr1997) [inline backbones/resnet.py:223]: 关于 deAtten2~5 的修改，可以去访问 `self.en_deconv` 参数，用来判断是否需要反卷积。比如这样：
```python
def forward(self, x):
    ......
    deAtten2 = self.upsample2(atten) if self.en_deconv else None # deConv for ER LOSS
    ......
    if self.en_deconv:
        return output, deAtten2, deAtten3, deAtten4, deAtten5  # used for ER loss
    else:
        return output
```

File: build.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import os
```
Comment (ygtxr1997) [inline build.py:1]: 关于原版 deformableconv 的引用，之前已经提到过。
我们不需要把 deformableconv 中所有的代码搬过来，而是在合适的位置加上 readme，说明怎么编译和使用即可。
根据我的理解，以下代码都是不需要放进我们的仓库的：
```python
build.py
src/*
make.sh
```

File: utils/transforms.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
+from __future__ import absolute_import
```
Comment (ygtxr1997) [inline utils/transforms.py:1]: Pytorch 有内置版本的 random erasing，所以就不需要把他的源码再搬过来了。精度应该是一致的。

File: train.py [hunk 12]
```diff
@@ -122,6 +126,16 @@ def main(args):
 
             with torch.cuda.amp.autocast(cfg.fp16):
                 final_pred = backbone(img)
+                #TODO：how to update the ER LOSS
```
Comment (ygtxr1997) [inline train.py:148]: 建议在 config.py 中加上是否开启 ER loss 的设置选项，例如 `config.en_erloss = False`。调用如下：
```python
if not cfg.en_erloss:
    final_pred = backbone(img)
else:
    final_pred, d1, d2, d3, d4 = model(img)
```
注意空格和变量名。

File: functions/__init__.py [hunk 13]
```diff
@@ -0,0 +1 @@
+from .deform_conv import DeformConvFunction, deform_conv_function
```
Comment (ygtxr1997) [inline functions/__init__.py:1]: 为什么 functions/ 和 modules/ 下面各有一个 deform_conv.py？这两个 deform_conv.py，我猜测一个是官方原版，一个是我们的改进版？
如果是这样，那直接把 deform_conv.py 分别命名成 deform_conv_vanilla.py 和 deform_conv_3d.py，并且直接放在 backbones 目录下即可：
```python
- backbones
  |--- resnet.py
  |--- cbam.py
  |--- deform_conv_vanilla.py
  |--- deform_conv_3d.py
```
